core-backend/wasmi/src/funcs/mod.rs

```diff
@@ -0,0 +1,1650 @@
+                    return Err(DummyHostError.into());
+                }
+            };
+
+            if last_idx > message.len() {
+                host_state.err = FuncError::ReadWrongRange(at..last_idx, message.len());
+                return Err(DummyHostError.into());
+            }
+
+            let buffer = message[at..last_idx].to_vec();
+            let mut memory_wrap = get_caller_memory(&mut caller, &memory);
+            match memory_wrap.write(destination_ptr, &buffer) {
+                Ok(_) => Ok(()),
+                Err(e) => {
+                    host_state_mut!(caller).err = e.into();
+
+                    Err(DummyHostError.into())
+                }
+            }
+        };
+
+        Func::wrap(store, func)
+    }
+
+    pub fn size(store: &mut Store<HostState<E>>, forbidden: bool) -> Func {
+        let func = move |mut caller: wasmi::Caller<'_, HostState<E>>| {
+            let host_state = host_state_mut!(caller);
+            if forbidden {
+                host_state.err = FuncError::Core(E::Error::forbidden_function());
+                return Err(DummyHostError.into());
+            }
+
+            let size = host_state.ext.size();
+            match size {
+                Ok(size) => match u32::try_from(size) {
+                    Ok(size) => Ok((size,)),
+                    Err(_) => {
+                        host_state.err = FuncError::HostError;
+                        Err(DummyHostError.into())
+                    }
+                },
+                Err(e) => {
+                    host_state.err = FuncError::Core(e);
+                    Err(DummyHostError.into())
+                }
+            }
+        };
+
+        Func::wrap(store, func)
+    }
+
+    pub fn exit(store: &mut Store<HostState<E>>, forbidden: bool, memory: WasmiMemory) -> Func {
+        let func = move |mut caller: wasmi::Caller<'_, HostState<E>>,
+                         value_dest_ptr: i32|
+              -> Result<(), Trap> {
+            if forbidden {
+                host_state_mut!(caller).err = FuncError::Core(E::Error::forbidden_function());
+                return Err(DummyHostError.into());
+            }
+
+            let value_dest_ptr = value_dest_ptr as u32 as usize;
+
+            let read_result = {
+                let memory_wrap = get_caller_memory(&mut caller, &memory);
+                read_memory_as::<ProgramId>(&memory_wrap, value_dest_ptr)
+            };
+
+            host_state_mut!(caller).err = match read_result {
+                Ok(pid) => FuncError::Terminated(TerminationReason::Exit(pid)),
+                Err(e) => e.into(),
+            };
+
+            Err(DummyHostError.into())
+        };
+
+        Func::wrap(store, func)
+    }
+
+    pub fn exit_code(store: &mut Store<HostState<E>>, forbidden: bool) -> Func {
+        let func = move |mut caller: wasmi::Caller<'_, HostState<E>>| {
+            let host_state = host_state_mut!(caller);
+            if forbidden {
```

Comment:
Can we do macros for `forbidden if ` ? I think it's very suitable here

Reply:
I'm not sure actually but anyway `exit_if` introduced. Please check how it looks now.

Reply:
Yes look ok)